src/services/refactors/convertStringOrTemplateLiteral.ts

```diff
@@ -0,0 +1,183 @@
+        return arrayToTree(nodes.slice(1), binary);
+    }
+
+    function isStringConcatenationValid(node: Node): boolean {
+        const { containsString, areOperatorsValid } = treeToArray(node);
+        return containsString && areOperatorsValid;
+    }
+
+    function transformTreeToArray(node: Node): ReadonlyArray<Expression> {
+        return treeToArray(node).nodes;
+    }
+
+    function treeToArray(node: Node): { nodes: ReadonlyArray<Expression>, containsString: boolean, areOperatorsValid: boolean} {
+        if (isBinaryExpression(node)) {
+            const { nodes: leftNodes, containsString: leftHasString, areOperatorsValid: leftOperatorValid } = treeToArray(node.left);
+            const { nodes: rightNodes, containsString: rightHasString, areOperatorsValid: rightOperatorValid } = treeToArray(node.right);
+
+            if (!leftHasString && !rightHasString) {
+                return { nodes: [node], containsString: false, areOperatorsValid: true };
+            }
+
+            const nodeOperatorValid = node.operatorToken.kind === SyntaxKind.PlusToken;
+            const isPlus = leftOperatorValid && nodeOperatorValid && rightOperatorValid;
+
+            return { nodes: leftNodes.concat(rightNodes), containsString: true, areOperatorsValid: isPlus };
+        }
+
+        return { nodes: [node as Expression], containsString: isStringLiteral(node), areOperatorsValid: true };
+    }
+
+    function createHead(nodes: ReadonlyArray<Expression>): [number, TemplateHead] {
+        let begin = 0;
+        const head = createTemplateHead("");
+
+        while (begin < nodes.length && isStringLiteral(nodes[begin])) {
+            const next = nodes[begin] as StringLiteral;
+            head.text = head.text + next.text;
+            begin++;
+        }
+
+        head.text = escapeText(head.text);
+        return [begin, head];
+    }
+
+    function nodesToTemplate(nodes: ReadonlyArray<Expression>) {
+        const templateSpans: TemplateSpan[] = [];
+        const [begin, head] = createHead(nodes);
+
+        if (begin === nodes.length) {
+            return createNoSubstitutionTemplateLiteral(head.text);
+        }
+
+        for (let i = begin; i < nodes.length; i++) {
+            let current = nodes[i];
+            let text = "";
+
+            while (i + 1 < nodes.length && isStringLiteral(nodes[i + 1])) {
+                const next = nodes[i + 1] as StringLiteral;
+                text = text + next.text;
+                i++;
+            }
+
+            text = escapeText(text);
+            const templatePart = i === nodes.length - 1 ? createTemplateTail(text) : createTemplateMiddle(text);
+
+            if (isParenthesizedExpression(current)) current = current.expression;
+            templateSpans.push(createTemplateSpan(current, templatePart));
+        }
+
+        return createTemplateExpression(head, templateSpans);
+    }
+
+    function escapeText(content: string) {
+        return content.replace("`", "\`")

```

Comment:
thank you, I missed that one
@ajafff  Do you know by chance how to get raw representation of normal string (not of template literal)?

Since escape sequences are already interpreted, it's not possible to catch with regex.
```ts
const ex7 = "Unicode \u0023 \u{0023} " + "Hex \x23 " + " Octal \43";
// Unicode # # Hex # Octal # 
const transformed7 = ex7.replace(/\\([0-7]+)/g, (_whole, n) =>"eee");
// Unicode # # Hex # Octal # 
// Therefore, regex fails
```

With String.raw it's possible to get raw representation
```ts
const rawVersion = String.raw`Unicode \u0023 \u{0023} Hex \x23  Octal \43`;
// Unicode \u0023 \u{0023} Hex \x23  Octal \43
```

But if I use String.raw as function, it behaves differently. 
```ts
const ex9 = ["Unicode \u0023 \u{0023} Hex \x23  Octal \43"]
const transformed9 = String.raw({ raw: ex9, ...ex9 });
// Unicode # # Hex #  Octal 43
```

Reply:
I found a solution but I am not sure if it's the right way to solve it.